File: pypy/sum_floats.py
"""Sum a list of floats using an explicit Python loop."""


def sum_floats(values):
    total = 0.0
    for v in values:
        total += v
    return total


def nansum_floats(values):
    # nansum semantics: treat NaN as 0 (skip it). `v != v` is true only for
    # NaN, so it avoids a math.isnan() call in the hot loop.
    total = 0.0
    for v in values:
        if v != v:
            continue
        total += v
    return total

File: pypy/c/sum_floats.c
#include "sum_floats.h"

double sum_floats(const double *values, size_t n) {
    double total = 0.0;
    for (size_t i = 0; i < n; i++) {
        total += values[i];
    }
    return total;
}

double nansum_floats(const double *values, size_t n) {
    /* nansum semantics: skip NaN. (x != x) is true only for NaN. */
    double total = 0.0;
    for (size_t i = 0; i < n; i++) {
        double x = values[i];
        if (x != x) continue;
        total += x;
    }
    return total;
}
